cdn/signed-urls/src/main/java/com/google/cdn/SignedCookies.java

```diff
@@ -0,0 +1,103 @@
+import java.util.Base64;
+import java.util.Calendar;
+import java.util.Date;
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+public class SignedCookies {
+
+  public static void main(String[] args) throws Exception {
+    // TODO(developer): Replace these variables before running the sample.
+
+    // The name of the signing key added to the back end bucket or service
+    String keyName = "YOUR-KEY-NAME";
+    // Path to the url signing key uploaded to the backend service/bucket, as a 16-byte array
+    String keyPath = "/path/to/key";
+    // The date that the signed URL expires
+    Date expirationTime = getTomorrow();
+    // URL of request
+    String requestUrl = "https://media.example.com/videos/id/main.m3u8?userID=abc123&starting_profile=1";
+    // URL prefix to sign as a string. urlPrefix must start with either http:// or
+    // https:// and should not include query parameters
+    String urlPrefix = "https://media.example.com/videos/";
+
+    // Read the key as a base64 url-safe encoded string, then convert to byte array
+    String base64String = new String(Files.readAllBytes(Paths.get(keyPath)),
+        StandardCharsets.UTF_8);
+    byte[] keyBytes = Base64.getUrlDecoder().decode(base64String);
+
+    // Create signed cookie from policy
+    String signedCookie = signCookie(urlPrefix, keyBytes, keyName, expirationTime);
+    System.out.println(signedCookie);
+  }
+
+  // Gets Date representation for tomorrow from Calendar
+  private static Date getTomorrow() {
+    Calendar cal = Calendar.getInstance();
+    cal.setTime(new Date());
+    cal.add(Calendar.DATE, 1);
+    return cal.getTime();
+  }
+
+  // Creates a signed cookie for the specified policy
+  public static String signCookie(String urlPrefix, byte[] key, String keyName,
+      Date expirationTime)
+      throws InvalidKeyException, NoSuchAlgorithmException {
+
+    if (urlPrefix.contains("?") || urlPrefix.contains("#")) {
+      throw new IllegalArgumentException("urlPrefix must not include query params: " + urlPrefix);
+    }
+    if (!urlPrefix.startsWith("http://") && !urlPrefix.startsWith("https://")) {
+      throw new IllegalArgumentException(
+          "urlPrefix must start with either http:// or https://: " + urlPrefix);
+    }
+
+    final long unixTime = expirationTime.getTime() / 1000;
+    String encodedUrlPrefix = Base64.getUrlEncoder().encodeToString(urlPrefix.getBytes(
+        StandardCharsets.UTF_8));
+    String policyToSign = String.format("URLPrefix=%s:Expires=%d:KeyName=%s", encodedUrlPrefix,
+        unixTime, keyName);
+
+    String signature = getSignatureForUrl(key, policyToSign);
+    return String.format("Cloud-CDN-Cookie=%s:Signature=%s", policyToSign, signature);
+  }
+
+  // Creates signature for input string with private key
+  private static String getSignatureForUrl(byte[] privateKey, String input)
+      throws InvalidKeyException, NoSuchAlgorithmException {
+
+    final String algorithm = "HmacSHA1";
```

Comment:
It seems that [recommended][1] algorithm is HMAC-SHA256. What is a reason to use "HmacSHA1"?

[1]: https://cloud.google.com/kms/docs/algorithms#algorithm_recommendations

Reply:
Thanks for pointing to the doc. No particular reason, I was simply referencing existing samples for [SignedUrls](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/cdn/signed-urls/src/main/java/com/google/cdn/SignedUrls.java#L69) and other languages ([python](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/HEAD/cdn/snippets.py#L138), [Go](https://github.com/GoogleCloudPlatform/golang-samples/blob/cba0742a460ff7738709ac8c3513a0359e75bb26/cdn/signedcookies/sign_cookie.go#LL44C14-L44C14))

Reply:
updated to HMAC-SHA256 in https://github.com/GoogleCloudPlatform/java-docs-samples/pull/8136/commits/27c959a62fadeaca73e1e985efefb91dc35cfb0e

Reply:
Circling back on this:
- Noticed that documentation page for signed_url is guiding people to use HMAC-SHA1 ([doc link](https://cloud.google.com/cdn/docs/using-signed-urls#custom-code:~:text=with%20HMAC%2DSHA1%20by%20using%20the%20secret%20key))
- Other language samples for [signed_cookies](https://cloud.google.com/cdn/docs/using-signed-cookies#programmatically_creating_signed_cookies) and [signed_url](https://cloud.google.com/cdn/docs/using-signed-urls#programmatically_create_signed_urls) are all using HMAC-SHA1.

Is the recommendation to use HMAC-SHA256 a more recent change? Or is there other explanation to this discrepancy?

Reply:
SHA1 is less resistant to brute-force attacks than SHA256. The [doc][1] with SHA1 recommendation was last modified in 2020. The [recommendation][2] for SHA256 was updated in 2022.

I cannot comment on other services. However, it is straightforward that we should recommend using more secure algorithm.

[2]: https://cloud.google.com/kms/docs/algorithms#algorithm_recommendations
[1]: https://cloud.google.com/cdn/docs/using-signed-urls#custom-code:~:text=with%20HMAC%2DSHA1%20by%20using%20the%20secret%20key

Reply:
I don't believe this suggestion was correct. These code samples are specifically about implementing clients for a specific GCP service that expects signatures to be performed with HMAC-SHA1.